### PR TITLE
[SPARK-20156][SQL][ML][FOLLOW-UP] Java String toLowerCase with Locale.ROOT

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?

Add `Locale.ROOT` to all internal calls to String `toLowerCase`, `toUpperCase`

## How was this patch tested?
existing tests

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -411,7 +412,7 @@ public UTF8String toUpperCase() {
   }
 
   private UTF8String toUpperCaseSlow() {
-    return fromString(toString().toUpperCase());
+    return fromString(toString().toUpperCase(Locale.ROOT));
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: I think we explicitly didn't change this on purpose; the point of fixing Locale.ROOT is to make sure that strings that aren't really user data that could well be locale-dependent don't vary. For example internal identifiers for compression types or impurity types.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala [hunk 2]
```diff
@@ -37,7 +39,7 @@ class Tokenizer @Since("1.4.0") (@Since("1.4.0") override val uid: String)
 
   override protected def createTransformFunc: String => Seq[String] = {
     // scalastyle:off caselocale
-    _.toLowerCase.split("\\s")
+    _.toLowerCase(Locale.ROOT).split("\\s")
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Tokenizer.scala]: Same here, this needs to be locale-aware. It's modifying user data.

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -650,10 +651,13 @@ object Assembly {
     },
     jarName in (Test, assembly) := s"${moduleName.value}-test-${version.value}.jar",
     mergeStrategy in assembly := {
-      case m if m.toLowerCase.endsWith("manifest.mf")          => MergeStrategy.discard
-      case m if m.toLowerCase.matches("meta-inf.*\\.sf$")      => MergeStrategy.discard
+      case m if m.toLowerCase(Locale.ROOT).endsWith("manifest.mf")
```
Comment (srowen) [inline project/SparkBuild.scala]: I agree with these.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVHeaderChecker.scala [hunk 4]
```diff
@@ -67,8 +69,8 @@ class CSVHeaderChecker(
           var (nameInSchema, nameInHeader) = (fieldNames(i), columnNames(i))
           if (!caseSensitive) {
             // scalastyle:off caselocale
-            nameInSchema = nameInSchema.toLowerCase
-            nameInHeader = nameInHeader.toLowerCase
+            nameInSchema = nameInSchema.toLowerCase(Locale.ROOT)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVHeaderChecker.scala]: These may be user-supplied column names so I think it has to be locale-sensitive

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala [hunk 5]
```diff
@@ -74,7 +76,7 @@ case class ResolveLambdaVariables(conf: SQLConf) extends Rule[LogicalPlan] {
   private val canonicalizer = {
     if (!conf.caseSensitiveAnalysis) {
       // scalastyle:off caselocale
-      s: String => s.toLowerCase
+      s: String => s.toLowerCase(Locale.ROOT)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala]: Not sure about this one. It's affecting variable names in a user's SQL? then it shouldn't change

**Review by srowen — CHANGES_REQUESTED** [review]
I didn't finish reviewing them all, but I think almost all of these were left locale-sensitive on purpose.

**Comment by zhengruifeng** [comment]
@srowen  Yes, we should keep user input data and column names. Thanks for your explain!

**Review by srowen — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
retest this please

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**